sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -138,46 +138,80 @@ case class Not(child: Expression)
 case class In(value: Expression, list: Seq[Expression]) extends Predicate {
 
   require(list != null, "list should not be null")
+
+  lazy val valExprs = value match {
+    case cns: CreateNamedStruct => cns.valExprs
+    case expr => Seq(expr)
+  }
+
+  override lazy val resolved: Boolean = {
+    lazy val checkForInSubquery = list match {
+      case (l @ ListQuery(sub, children, _)) :: Nil =>
+        // SPARK-21759:
+        // It is possibly that the subquery plan has more output than value expressions, because
+        // the condition expressions in `ListQuery` might use part of subquery plan's output.
```

Comment:
So we are adding another criteria to consider an in-subquery expression to be resolved. That new criteria is - 
1) Any additional output attributes that may have been added to the subquery plan by optimizer should have a reference in the originating in-subquery expression's children.(children reflect the pulled up correlated predicates)

Just for my understanding, there is no way to trigger this condition from our regular code path, right ? This is just to guard against any potential incorrect rewrites by the optimizer in the future ?